### PR TITLE
Fix attach volume to instance repeatedly

### PR DESCRIPTION
**Pull request by FengyunPan**
1.When volume's status is 'attaching', controllermanager will attach
    it again and return err. So it is necessary to check volume's
    status before attach/detach volume.

   2. When volume's status is 'attaching', its attachments will be None,
    controllermanager can't get device path and make some failed event.
    But it is normal, so don't return err when attachments is None

Fix bug: #44536

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
is there a test case that can demonstrate the problem that was fixed?

**Comment by FengyunPan** [comment]
@derekwaynecarr Hi, I have added a test case for it.

File: pkg/volume/cinder/attacher.go [hunk 1]
```diff
@@ -72,6 +72,14 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
 	if err != nil {
 		return "", err
 	}
+	HasDone, VolumeStatus, err := attacher.cinderProvider.OperateHasDone(volumeID)
```
Comment (gnufied) [inline pkg/volume/cinder/attacher.go]: Fix the naming of these variables please. Make sure they start with lower case letters.

File: pkg/volume/cinder/attacher.go [hunk 2]
```diff
@@ -89,18 +97,27 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
...
+			glog.Infof("Attach volume %q to instance %q failed with: %v", volumeID, instanceid, err)
 			return "", err
 		}
 	}
 
-	devicePath, err := attacher.cinderProvider.GetAttachmentDiskPath(instanceid, volumeID)
+	// When volume's status is 'attaching', DiskIsAttached() return <false, nil>, and can't get device path.
+	// We should get device path next time and do not return err.
+	devicePath := ""
+	HasDone, _, err = attacher.cinderProvider.OperateHasDone(volumeID)
```
Comment (gnufied) [inline pkg/volume/cinder/attacher.go]: same as before about naming.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 3]
```diff
@@ -201,12 +206,34 @@ func (volumes *VolumesV2) deleteVolume(volumeName string) error {
 	return err
 }
 
+// return true when volume has been attached or detached.
+// or return false.
+func (os *OpenStack) OperateHasDone(diskName string) (bool, string, error) {
```
Comment (gnufied) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Can you rename this function to something like - `OperationPending` ?

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 4]
```diff
@@ -74,8 +74,14 @@ type VolumeCreateOpts struct {
 	Metadata     map[string]string
 }
 
-func (volumes *VolumesV1) createVolume(opts VolumeCreateOpts) (string, error) {
+const (
+	volumeAvailableStatus = "available"
+	volumeInUseStatus = "in-use"
+	VolumeDeletedStatus = "deleted"
+	VolumeErrorStatus = "error"
+)
```
Comment (gnufied) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Lets be consistent about naming. Constants generally start with upper case letter.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 5]
```diff
@@ -201,12 +206,34 @@ func (volumes *VolumesV2) deleteVolume(volumeName string) error {
...
+// return true when volume has been attached or detached.
+// or return false.
+func (os *OpenStack) OperateHasDone(diskName string) (bool, string, error) {
+	volume, err := os.getVolume(diskName)
+	if err != nil {
+		return false, "", err
+	}
+	VolumeStatus := volume.Status
+	if (VolumeStatus == VolumeErrorStatus) {
+		glog.Errorf("volume's status is %s", VolumeStatus)
```
Comment (gnufied) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Can you log the disk name when we are logging the error? Without disk name, this error log will be probably useless

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 6]
```diff
@@ -201,12 +206,34 @@ func (volumes *VolumesV2) deleteVolume(volumeName string) error {
...
+}
+
 // Attaches given cinder volume to the compute running kubelet
 func (os *OpenStack) AttachDisk(instanceID string, diskName string) (string, error) {
 	volume, err := os.getVolume(diskName)
 	if err != nil {
 		return "", err
 	}
+	if volume.Status != volumeAvailableStatus {
+		errmsg := fmt.Sprintf("volume %s status is %s, not %s, can not be attached to instance %s.", volume.Name, volume.Status, volumeAvailableStatus, instanceID)
```
Comment (gnufied) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Please use ` fmt.Errorf`

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 7]
```diff
@@ -378,7 +413,7 @@ func (os *OpenStack) GetAttachmentDiskPath(instanceID string, diskName string) (
 			return "", errors.New(errMsg)
 		}
 	}
-	return "", fmt.Errorf("volume %s is not attached to %s", diskName, instanceID)
+	return "", fmt.Errorf("volume %s has losed ServerId.", diskName)
 }
```
Comment (gnufied) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: typo?

**Review by gnufied — CHANGES_REQUESTED** [review]

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 4 above)
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Done.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 5 above)
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Done.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 3 above)
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Done.

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 6 above)
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: the errmsg of errors.New(errmsg) should be a string not a error

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go (hunk 7 above)
Comment (FengyunPan) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: Done

File: pkg/volume/cinder/attacher.go (hunk 1 above)
Comment (FengyunPan) [inline pkg/volume/cinder/attacher.go]: Done

File: pkg/volume/cinder/attacher.go (hunk 2 above)
Comment (FengyunPan) [inline pkg/volume/cinder/attacher.go]: Done

Comment (FengyunPan) [inline pkg/volume/cinder/attacher.go]: Have updated it.

**Comment by idvoretskyi** [comment]
cc @kubernetes/sig-openstack-pr-reviews

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by dims** [comment]
@FengyunPan Can  you please check on the failed jobs?

**Comment by FengyunPan** [comment]
@k8s-bot bazel test this
@k8s-bot unit test this

**Comment by FengyunPan** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by FengyunPan** [comment]
@k8s-bot kops aws e2e test this

**Comment by FengyunPan** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by FengyunPan** [comment]
@dims I have checked the failed jobs, we can ignore "pull-kubernetes-federation-e2e-gce" jobs because it is federation's jobs.

**Comment by dims** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by FengyunPan** [comment]
@jsafrane  @mikedanese Hi,  I'd like you to do a code review, thank you.

**Comment by FengyunPan** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by FengyunPan** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by saad-ali** [comment]
This PR broke Flex volumes. See https://github.com/kubernetes/kubernetes/issues/46882#issuecomment-306331570. Proposing reverting (at least partially this PR). Please discuss on that issue.

**Pull request merged**